src/sentry/taskworker/worker.py

```diff
@@ -158,7 +167,18 @@ def process_task(self, activation: TaskActivation) -> TaskActivation | None:
                 status=TASK_ACTIVATION_STATUS_FAILURE,
             )
 
-        # TODO(taskworker): Add at_most_once checks
+        if activation.retry_state.at_most_once:
+            key = get_at_most_once_key(activation.namespace, activation.taskname, activation.id)
+            if cache.add(key, "1", timeout=AT_MOST_ONCE_TIMEOUT):  # The key didn't exist
+                metrics.incr(
+                    "taskworker.task.at_most_once.executed", extra={"task": activation.taskname}
+                )
+            else:
+                metrics.incr(
+                    "taskworker.worker.at_most_once.skipped", extra={"task": activation.taskname}
+                )
+                return None
```

Comment:
We could reply to the broker here that the task is 'complete'. That could help prevent the task from being given to another worker as this worker would appear to be 'dead' due to no response.

Reply:
True, the task could be running in another worker. The scenario I was concerned about is:

1. Worker A picks up an at_most_once task. The task will be status=processing
2. Worker A dies and can't send an update to taskbroker.
3. The task will exceed its processing deadline, and be put back into pending.
4. Worker B takes the task and returns early without any updates. The task will once again exceed its processing_deadline, and we'll burn worker time looping from steps 2 to 4 until the message is deadlettered.

We could have another status/state for 'failed because of idempotency' that could escape the loop. I haven't thought through how else that status should be materially different from `failure` though.

Reply:
What if the worker marked the task completed as soon as it added the task to the idempotency cache? There's no way for the broker to safely assign the task again after it has been assigned once. And if the task is failing before/during setting the ID, we know that know actual work has been done so another worker could still execute it.

Reply:
The alternative to this is to propagate the at_most_once state into the protobuf. Then the broker will know to never retry an at most once task, and it can send it immediately to deadletter queue on a processing deadline.

Reply:
Including `at_most_once` in the protobuf would let us solve processing_deadlines more efficiently as the broker could skip doing deadline retries if the worker never responds, and regular retries could continue to work correctly, despite retries + at_most_once being nonsense.

Reply:
Given that, I think this PR is OK to be merged as is. If a worker happens to get assigned a task that is in the cache, it should skip it.

There will be a followup PR in the taskbroker to do some checks to avoid that scenario.